projects/openexr/openexr_deeptiles_fuzzer.cc

```diff
@@ -0,0 +1,196 @@
+
+  for (int i = 0; i < channelCount; i++) {
+    data[i].resizeErase(height, width);
+  }
+
+  DeepFrameBuffer frameBuffer;
+
+  int memOffset = dataWindow.min.x + dataWindow.min.y * width;
+  frameBuffer.insertSampleCountSlice(
+      Slice(IMF::UINT, (char *)(&localSampleCount[0][0] - memOffset),
+            sizeof(unsigned int) * 1, sizeof(unsigned int) * width));
+
+  for (int i = 0; i < channelCount; i++) {
+    std::stringstream ss;
+    ss << i;
+    std::string str = ss.str();
+
+    int sampleSize = sizeof(float);
+
+    int pointerSize = sizeof(char *);
+
+    frameBuffer.insert(
+        str, DeepSlice(IMF::FLOAT, (char *)(&data[i][0][0] - memOffset),
+                       pointerSize * 1, pointerSize * width, sampleSize));
+  }
+
+  part->setFrameBuffer(frameBuffer);
+  for (int ly = 0; ly < part->numYLevels(); ly++) {
+    for (int lx = 0; lx < part->numXLevels(); lx++) {
+      Box2i dataWindowL = part->dataWindowForLevel(lx, ly);
+
+      part->readPixelSampleCounts(0, part->numXTiles(lx) - 1, 0,
+                                  part->numYTiles(ly) - 1, lx, ly);
+
+      for (int i = 0; i < part->numYTiles(ly); i++) {
+        for (int j = 0; j < part->numXTiles(lx); j++) {
+          Box2i box = part->dataWindowForTile(j, i, lx, ly);
+          for (int y = box.min.y; y <= box.max.y; y++)
+            for (int x = box.min.x; x <= box.max.x; x++) {
+              int dwy = y - dataWindowL.min.y;
+              int dwx = x - dataWindowL.min.x;
+
+              for (int k = 0; k < channelCount; k++) {
+                data[k][dwy][dwx] = new float[localSampleCount[dwy][dwx]];
+              }
+            }
+        }
+      }
+
+      try {
+        part->readTiles(0, part->numXTiles(lx) - 1, 0, part->numYTiles(ly) - 1,
+                        lx, ly);
+      } catch (...) {
+      }
+
+      for (int i = 0; i < part->levelHeight(ly); i++) {
+        for (int j = 0; j < part->levelWidth(lx); j++) {
+          for (int k = 0; k < channelCount; k++) {
+            delete[](float *) data[k][i][j];
+          }
+        }
+      }
+    }
+  }
+}
+
+static void readFileSingle(const char filename[]) {
+  DeepTiledInputFile *file;
+  try {
+    file = new DeepTiledInputFile(filename, 8);
+  } catch (...) {
+    return;
+  }
+
+  try {
+    readFile(file);
+  } catch (std::exception &e) {
+  }
+
+  delete file;
+}
+
+static void readFileMulti(const char filename[]) {
+  MultiPartInputFile *file;
+
+  try {
+    file = new MultiPartInputFile(filename, 8);
+  } catch (...) {
+    return;
+  }
+
+  for (int p = 0; p < file->parts(); p++) {
+    DeepTiledInputPart *part;
+    try {
+      part = new DeepTiledInputPart(*file, p);
+    } catch (...) {
+      continue;
+    }
+
+    try {
+      readFile(part);
+    } catch (...) {
+    }
+
+    delete part;
+  }
+
+  delete file;
+}
+
+}  // namespace
+
+// from cl/164883104
+static char *buf_to_file(const char *buf, size_t size) {
+  char *name = strdup("/dev/shm/fuzz-XXXXXX");
```

Comment:
this isn't recommended. Instead we suggest to use an in memory file if possible

Reply:
@rjotwani and I are working on a second PR that modifies these fuzzers to use OpenEXR's `Imf::IStream` class for in-memory input.  We can merge that PR with this one if you'd prefer that over reviewing it separately.